### DISCUSSION TITLE
Discrepancy in Training Outcomes: `Problem` Class with `Trainer` vs. PenaltyLoss with Custom Training Loop for extended Linear Regression example with Parameter Constraints

### DISCUSSION THREAD

**Discussion by ninadkgaikwad** [pnnl/neuromancer#226] [Q&A]
I am working on a constrained optimization problem using **Neuromancer**, and I have implemented two versions of the training setup: (extending the linear regression example to include constraints over the parameters [w, b] to understand Neuromancer a bit better for some custom NLP implementations to follow later on)

1. **With the `Problem` class**: Uses `Trainer` for training.  
2. **Without the `Problem` class**: Uses a custom training loop.  

Despite keeping **the same data, loss function, constraints, and optimizer settings**, the **learned parameters (`w_2`, `b_2`) and loss trajectories differ** significantly between these two approaches. Moreover, the results from the custom training loop without using the Problem class are approximately equal to true parameters, while those from the Trainer + Problem class do not come close to the true parameters. 

It seems like two different losses are being used for backpropagation in the two implementations (this is not intended). One thing to note is that in the case where the Problem class is used, the component list used is defined as functions without any trainable parameters so that the component can be used to pass data just to the PenaltyLoss. 

Can someone please help me understand why this discrepancy occurs and identify any mistakes in my formulation, implementation, or conceptual understanding?

I am attaching the code for both the implementations here:
---

## Code with `Problem` Class (Using `Trainer`)  

```python
from neuromancer.constraint import variable
import torch
import torch.nn as nn
import torch.nn.functional as F
import neuromancer as nm
from neuromancer.trainer import Trainer
from neuromancer.problem import Problem
from neuromancer.constraint import variable
from neuromancer.dataset import DictDataset
from neuromancer.loss import PenaltyLoss
from neuromancer.modules import blocks
from neuromancer.system import Node

"""### Data"""

# creating total data points
nsim =  10000

w_true_value = 2.5
b_true_value = 1.5

x_min = -100
x_max = 100

#Training Hyper Parameters
Test_split = 0.7
Dev_split = 0.2

lr = 0.001      # step size for gradient descent
epochs = 10    # number of training epochs
warmup = 100    # number of epochs to wait before enacting early stopping policy
patience = 100  # number of epochs with no improvement in eval metric to allow before early stopping

# Setting values for batch size, number of workers and shuffle
batch_size_value=32
num_workers_value=0,
shuffle_value=True

# Basic Computation
step = (x_max-x_min) / nsim

# randomly generate the problem data
torch.manual_seed(0)
x_true = torch.arange(x_min, x_max+step, step)
w_true = torch.tensor(w_true_value)
b_true = torch.tensor(b_true_value)
y_true = x_true * w_true + b_true

#Create train, dev and test split data

# Calculate split indices
num_test = int(nsim * Test_split)
num_remaining = nsim - num_test
num_dev = int(num_remaining * Dev_split)
num_train = num_remaining - num_dev

# Split the data
x_test, y_test = x_true[:num_test], y_true[:num_test]
x_dev, y_dev = x_true[num_test:num_test + num_dev], y_true[num_test:num_test + num_dev]
x_train, y_train = x_true[num_test + num_dev:], y_true[num_test + num_dev:]

x_test = x_test.reshape(x_test.shape[0], 1)
y_test = x_test.reshape(y_test.shape[0], 1)
x_dev = x_dev.reshape(x_dev.shape[0], 1)
y_dev = x_dev.reshape(y_dev.shape[0], 1)
x_train = x_train.reshape(x_train.shape[0], 1)
y_train = x_train.reshape(y_train.shape[0], 1)

# create dictionaries with sampled datapoints 
samples_train = {"x" : x_train,
               "y" : y_train}
samples_dev = {"x" : x_dev,
               "y" : y_dev}
samples_test = {"x" : x_test,
               "y" : y_test}

# create named dictionary datasets
train_data = DictDataset(samples_train, name='train')
dev_data = DictDataset(samples_dev, name='dev')
test_data = DictDataset(samples_test, name='test')

# create torch dataloaders for the Trainer
train_loader = torch.utils.data.DataLoader(train_data, batch_size=batch_size_value, num_workers=0,
                                           collate_fn=train_data.collate_fn, shuffle=shuffle_value)
dev_loader = torch.utils.data.DataLoader(dev_data, batch_size=batch_size_value, num_workers=0,
                                           collate_fn=dev_data.collate_fn, shuffle=shuffle_value)
test_loader = torch.utils.data.DataLoader(test_data, batch_size=batch_size_value, num_workers=0,
                                           collate_fn=test_data.collate_fn, shuffle=shuffle_value)



"""### Neuromancer variable model"""

# randomly initialize parameters
torch.manual_seed(0)
#w_2 = torch.randn(1, requires_grad=True)
#b_2 = torch.randn(1, requires_grad=True)

w_2 = torch.tensor([2.0], dtype=torch.float32, requires_grad=True)
b_2 = torch.tensor([1.1], dtype=torch.float32, requires_grad=True)

# create symbolic representation of the problem using Neuromancer variable
var_x = variable("x1")
var_w = variable(w_2,  display_name='w_2')
var_b = variable(b_2,  display_name='b_2')
var_y_est = var_x * var_w + var_b
var_y_true = variable("y1")
var_loss = F.mse_loss(var_y_est, var_y_true)

fun_1 = lambda x1: x1
node_1 = Node(fun_1, ['x'], ['x1'], name='quadratic') 

fun_2 = lambda x1: x1
node_2 = Node(fun_2, ['y'], ['y1'], name='quadratic1') 

obj = var_loss.minimize(weight=1.0, name='obj')
con_1 = 1.* var_w >= 1
con_2 = 1.* var_b >= 1
con_1.name = 'c1'
con_2.name = 'c2'

# constrained optimization problem construction
objectives = [obj]
constraints = [con_1, con_2]
components = [node_1, node_2]

# create penalty method loss function
loss = PenaltyLoss(objectives, constraints)
# construct constrained optimization problem
problem = Problem(components, loss)
#problem.show()

# training
optimizer = torch.optim.Adam(problem.parameters(), lr=lr)

# define trainer
trainer = Trainer(
    problem,
    train_loader,
    dev_loader,
    test_loader,
    optimizer,
    epochs=epochs,
    patience=patience,
    warmup=warmup,
    clip=torch.inf)


# Train NLP solution map
best_model = trainer.train()
best_outputs = trainer.test(best_model)
# load best model dict
problem.load_state_dict(best_model)    

print(f"w estimated to {w_2} true value is {w_true}")
print(f"b estimated to {b_2} true value is {b_true}")
```

---

## Code Without Problem Class (Custom Training Loop)

```python

from neuromancer.constraint import variable
import torch
import torch.nn.functional as F
import neuromancer as nm
from neuromancer.trainer import Trainer
from neuromancer.problem import Problem
from neuromancer.constraint import variable
from neuromancer.dataset import DictDataset
from neuromancer.loss import PenaltyLoss
from neuromancer.modules import blocks
from neuromancer.system import Node

"""### Data"""

# creating total data points
nsim =  10000

w_true_value = 2.5
b_true_value = 1.5

x_min = -100
x_max = 100

#Training Hyper Parameters
Test_split = 0.7
Dev_split = 0.2

lr = 0.001      # step size for gradient descent
epochs = 10    # number of training epochs
warmup = 100    # number of epochs to wait before enacting early stopping policy
patience = 100  # number of epochs with no improvement in eval metric to allow before early stopping

# Setting values for batch size, number of workers and shuffle
batch_size_value=32
num_workers_value=0,
shuffle_value=True

# Basic Computation
step = (x_max-x_min) / nsim

# randomly generate the problem data
torch.manual_seed(0)
x_true = torch.arange(x_min, x_max+step, step)
w_true = torch.tensor(w_true_value)
b_true = torch.tensor(b_true_value)
y_true = x_true * w_true + b_true

#Create train, dev and test split data
# Calculate split indices
num_test = int(nsim * Test_split)
num_remaining = nsim - num_test
num_dev = int(num_remaining * Dev_split)
num_train = num_remaining - num_dev

# Split the data
x_test, y_test = x_true[:num_test], y_true[:num_test]
x_dev, y_dev = x_true[num_test:num_test + num_dev], y_true[num_test:num_test + num_dev]
x_train, y_train = x_true[num_test + num_dev:], y_true[num_test + num_dev:]

# create dictionaries with sampled datapoints 
samples_train = {"x" : x_train,
               "y" : y_train}
samples_dev = {"x" : x_dev,
               "y" : y_dev}
samples_test = {"x" : x_test,
               "y" : y_test}

# create named dictionary datasets
train_data = DictDataset(samples_train, name='train')
dev_data = DictDataset(samples_dev, name='dev')
test_data = DictDataset(samples_test, name='test')

# create torch dataloaders for the Trainer
train_loader = torch.utils.data.DataLoader(train_data, batch_size=batch_size_value, num_workers=0,
                                           collate_fn=train_data.collate_fn, shuffle=shuffle_value)
dev_loader = torch.utils.data.DataLoader(dev_data, batch_size=batch_size_value, num_workers=0,
                                           collate_fn=dev_data.collate_fn, shuffle=shuffle_value)
test_loader = torch.utils.data.DataLoader(test_data, batch_size=batch_size_value, num_workers=0,
                                           collate_fn=test_data.collate_fn, shuffle=shuffle_value)

"""### Neuromancer variable model"""

#  initialize parameters
w_2 = torch.tensor([2.0], dtype=torch.float32, requires_grad=True)
b_2 = torch.tensor([1.1], dtype=torch.float32, requires_grad=True)

# create symbolic representation of the problem using Neuromancer variable
var_x = variable("x")
var_w = variable(w_2,  display_name='w_2')
var_b = variable(b_2,  display_name='b_2')
var_y_est = var_x * var_w + var_b
var_y_true = variable("y")
var_loss = F.mse_loss(var_y_est, var_y_true)

obj = var_loss.minimize(weight=1.0)
con_1 = 1.* var_w >= 1
con_2 = 1.* var_b >= 1

# constrained optimization problem construction
objectives = [obj]
constraints = [con_1, con_2]

# create penalty method loss function
loss = PenaltyLoss(objectives, constraints)

#For-Loop

# Initialize the optimizer with the parameters from the loss object
opt_var = torch.optim.Adam(loss.parameters(), lr=lr)

# Arrays to store loss values for analysis
train_losses = []
val_losses = []
best_val_loss = float('inf')
best_model = None

# Training loop with batch processing
for epoch in range(epochs):
    print(f"Epoch {epoch+1}/{epochs}")

    # Training
    total_train_loss = 0
    for batch in train_loader:
        opt_var.zero_grad()  # Zero the gradients
        # Construct input data for loss function
        #input_data = {'x': batch['x'], 'y': batch['y']}
        # Assuming the 'loss' function can directly take the batch dictionary and return a dict with 'loss'
        computed_loss = loss(batch)
        train_loss = computed_loss['loss']
        train_loss.backward()  # Backpropagate the error
        opt_var.step()  # Update parameters
        total_train_loss += train_loss.item()

    avg_train_loss = total_train_loss / len(train_loader)
    train_losses.append(avg_train_loss)
    print(f"Training Loss: {avg_train_loss:.4f}")

    # Validation
    total_val_loss = 0
    with torch.no_grad():  # No gradient tracking in validation
        for batch in dev_loader:
            #input_data = {'x': batch['x'], 'y': batch['y']}
            computed_loss = loss(batch)
            val_loss = computed_loss['loss']
            total_val_loss += val_loss.item()
    avg_val_loss = total_val_loss / len(dev_loader)
    val_losses.append(avg_val_loss)
    print(f"Validation Loss: {avg_val_loss:.4f}")


# Output the estimated parameters
print(f"Estimated w_2: {w_2}")
print(f"Estimated b_2: {b_2}")
```
